Add WP-CLI command to manage option autoload

trepmal opened the pull request:
## Description

Add command for managing option autoload. See #453

Also https://github.com/trepmal/option-autoload-cli

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [x] This change has relevant documentation additions / updates (if applicable). **has inline docs**

## Steps to Test

With PR applied, can test with:

```
wp option autoload get home # should be 'yes' at start
wp option autoload set home no
wp option autoload get home # should be 'no'
wp option autoload set home yes # restore
wp option autoload list no # should list starting with `moderation_keys`, `recently_edited`
```

mjangda commented:
@trepmal nice work!

Are there differences between this and the command on your repo? Ideally they'd be the same.

trepmal commented:
The only diffs between this and my repo are 
1) this version is condensed. This class and add_command line are in this one file. 
2) this uses WPCOM_VIP_CLI_Command

emrikol commented:
Cool!  This is only a temporary thing, right?  Once the option is `update_option()`'d again without autoload, will it revert to autoloading again?

WPprodigy commented:
> Cool! This is only a temporary thing, right? Once the option is update_option()'d again without autoload, will it revert to autoloading again?

Sort of. If the option isn't technically updated/changed, then the next `update_option()` won't make any difference to the autoloading - afaik.

But secondly, I'm pretty positive that the third `$autoload` arg is rarely used. If the code specifically asks for autoloading, then it makes sense to respect that. In cases where we need to flip the switch, it's quite likely there is no preference to autoloading in the code. <2¢ />

trepmal commented:
> Cool! This is only a temporary thing, right? Once the option is update_option()'d again without autoload, will it revert to autoloading again?

If the `autoload` arg is not set, then changing the value in the db will stick.
If `autoload` is set, it will be applied the next time the option is updated _with a new value_

```
$ wp option update TESTING test      # prep option
Success: Updated 'TESTING' option.

$ wp option autoload get TESTING     # default autoload
yes

$ wp option autoload set TESTING no  # change it
Success: Autoload changed. Cache flushed.

$ wp option update TESTING test2     # update value
Success: Updated 'TESTING' option.

$ wp option autoload get TESTING     # autoload unchanged
no

$ wp option update TESTING test3 --autoload=yes # update value with new autoload
Success: Updated 'TESTING' option.

$ wp option autoload get TESTING     # autoload changed
yes

$ wp option update TESTING test3 --autoload=no # same value with new autoload
Error: Could not update option 'TESTING'.

$ wp option autoload get TESTING     # autoload unchanged
yes
```

I feel like once upon a time `autoload` could only be set with `add_option()`? It seems like I rarely see the autoload arg with `update_option()`.

I think we'd only run into a self-unsolving issue is if something is forcibly setting autoload to 'yes' , which seems quite unlikely.

rogertheriault commented:
Are our style requirements that strict? Wow. https://circleci.com/gh/Automattic/vip-go-mu-plugins/3392

mjangda commented:
> Are our style requirements that strict?

We're just inheriting from the WordPress + VIP Go standards. Also open to tweaks on standards we don't agree with.

mjangda commented:
> The only diffs between this and my repo are

It would be good to minimize the differences so it's easier to port changes across. I think `1` is probably fine but I would drop `2` unless there are strong reasons for inheriting from the WPCOM class.

hanifn commented on wp-cli/vip-option-autoload.php:
```diff
@@ -0,0 +1,219 @@
+		}
+
+		WP_CLI::success( sprintf(
+			'Autoload changed.%s',
+			WP_CLI::colorize( $cache_success ? ' Cache flushed.' : ' %rCache flush failed.%n' )
```
The spacing looks a little odd here.

Maybe something like this is better?
```suggestion
			'Autoload changed. %s',
			WP_CLI::colorize( $cache_success ? 'Cache flushed.' : '%rCache flush failed.%n' )
```

brandon-m-skinner replied:
It's the difference between:

`Success: Autoload changed. Cache flushed.`

and

`Success: Autoload changed.Cache flushed.`

The space should stay for formatting purposes.

brandon-m-skinner replied:
It's needed for formatting. Otherwise there's no space between the cache statement and the rest of the output.

hanifn replied:
Yes but when using something like `sprintf()`, you would usually put the spaces in the format string instead of the parameters being inserted. That way, the formatting is handled by the formatting string alone and is not a concern when inserting the parameters. Still, it's minor matter.

brandon-m-skinner replied:
Whoops, sorry I'm blind 🤦‍♂

brandon-m-skinner commented:
@trepmal I just sorted out the lint issues since I've been dealing with them a lot lately. If this all looks good to you, we'll approve and deploy :+1:

trepmal commented on wp-cli/class-option-autoload.php:
```diff
@@ -0,0 +1,226 @@
+				'autoload' => $yn,
+				'fields' => 'option_name',
+				'format' => $assoc_args['format'],
+			],
```
This comma is valid, but I hate it :)

trepmal replied:
Well, it's only valid on PHP 7.3+ So I'm not keeping it upstream.

brandon-m-skinner replied:
SVN pre-commit hook didn't like it either. It's been removed in master.

trepmal commented:
I've ported most of the changes upstream and tests are passing there. See inline comment for info

brandon-m-skinner commented:
r868-stacks
